File: app/src/main/java/com/grohden/niceanimals/ui/holders/NAViewHolder.java
package com.grohden.niceanimals.ui.holders;

import android.content.Context;
import android.content.Intent;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.ImageView;
import android.widget.ProgressBar;

import com.grohden.niceanimals.R;
import com.grohden.niceanimals.realm.entities.NiceAnimal;
import com.grohden.niceanimals.ui.activities.GalleryActivity;
import com.squareup.picasso.Callback;
import com.squareup.picasso.Picasso;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

import static com.grohden.niceanimals.ui.activities.GalleryActivity.IMAGE_URL_EXTRA;

public class NAViewHolder extends RecyclerView.ViewHolder implements Callback {

    @BindView(R.id.pic_image_view)
    ImageView mAnimalView;

    @BindView(R.id.pic_progress_bar)
    ProgressBar mProgressBar;


    public NAViewHolder(View itemView) {
        super(itemView);
        ButterKnife.bind(this, itemView);

    }

    public void bindAnimal(NiceAnimal animal) {
        mProgressBar.setVisibility(View.VISIBLE);

        Picasso.get()
                .load(animal.getPictureUrl())
                .resizeDimen(R.dimen.min_image_width, R.dimen.min_image_height)
                .centerCrop()
                .into(mAnimalView, this);
    }

    @OnClick({R.id.pic_image_view})
    public void openFullscreenImageActivity(ImageView view) {
        final Context context = view.getContext();
        final Intent intent = new Intent(
                context,
                GalleryActivity.class
        );

        intent.putExtra(IMAGE_URL_EXTRA, getAdapterPosition());

        context.startActivity(intent);
    }


    @Override
    public void onSuccess() {
        mProgressBar.setVisibility(View.GONE);
    }

    @Override
    public void onError(Exception e) {
        mProgressBar.setVisibility(View.GONE);
    }
}

File: app/src/main/java/com/grohden/niceanimals/ui/holders/NAViewHolder.kt
package com.grohden.niceanimals.ui.holders

import android.support.v7.widget.RecyclerView
import android.view.View
import android.widget.ImageView
import android.widget.ProgressBar
import com.grohden.niceanimals.R
import com.grohden.niceanimals.realm.entities.NiceAnimal
import com.grohden.niceanimals.ui.activities.GalleryActivity
import com.squareup.picasso.Callback
import com.squareup.picasso.Picasso

class NAViewHolder(animalCard: View) : RecyclerView.ViewHolder(animalCard), Callback {

    private val progressBar: ProgressBar by lazy {
        animalCard.findViewById<ProgressBar>(R.id.pic_progress_bar)
    }

    private val animalView: ImageView by lazy {
        animalCard.findViewById<ImageView>(R.id.pic_image_view)
    }

    init {
        animalView.setOnClickListener(this::openFullscreenImageActivity)
    }

    fun bindAnimal(animal: NiceAnimal) {
        progressBar.visibility = View.VISIBLE

        Picasso.get()
                .load(animal.pictureUrl)
                .resizeDimen(R.dimen.min_image_width, R.dimen.min_image_height)
                .centerCrop()
                .into(animalView, this)
    }

    private fun openFullscreenImageActivity(view: View) {
        val context = view.context
        val intent = GalleryActivity.createIntent(
                context,
                adapterPosition
        )

        context.startActivity(intent)
    }


    override fun onSuccess() {
        progressBar.visibility = View.GONE
    }

    override fun onError(e: Exception) {
        progressBar.visibility = View.GONE
    }
}
